File: src/aetherius/core/events/sinks.py
"""Event sinks: LogSink for terminal output, NullSink for tests.

WebSocketSink is deferred to the daemon milestone.
"""

from __future__ import annotations

import logging
from typing import Protocol, runtime_checkable

from .models import EventType, RunEvent

_log = logging.getLogger("aetherius.events")


@runtime_checkable
class Sink(Protocol):
    def on_event(self, event: RunEvent) -> None: ...


class NullSink:
    """Discards all events. Used in tests."""

    def on_event(self, event: RunEvent) -> None:
        pass


class LogSink:
    """Writes events to Python logging.

    In debug mode every event is logged at DEBUG; otherwise only INFO and above.
    """

    def __init__(self, *, debug: bool = False) -> None:
        self._debug = debug

    def on_event(self, event: RunEvent) -> None:
        level_str = event.level or ("debug" if event.type == EventType.DEBUG else "info")
        level = getattr(logging, level_str.upper(), logging.INFO)
        if not self._debug and level < logging.INFO:
            return
        msg = f"[{event.type.value}]"
        if event.step_id:
            msg += f" step={event.step_id}"
        if event.message:
            msg += f" {event.message}"
        _log.log(level, msg)

File: sdks/engine/src/events/sinks.ts
/**
 * Ready-made sinks, mirror of `src/aetherius/core/events/sinks.py`.
 *
 * The engine ships only what it can honour anywhere: discarding, collecting, and handing events to
 * a callback. Rendering belongs to the host — a terminal logger would be dead weight in an app, and
 * the application's own UI is the real consumer (milestone 3-E).
 */

import type { RunEvent, RunEventHandler, Sink } from "./models.js";

/** Discards every event. Used in tests and when a caller wants no stream at all. */
export class NullSink implements Sink {
  onEvent(): void {
    /* intentionally empty */
  }
}

/** Keeps every event in order. The assertion surface of the test suite. */
export class CollectingSink implements Sink {
  readonly events: RunEvent[] = [];

  onEvent(event: RunEvent): void {
    this.events.push(event);
  }
}

/** Adapts a plain callback — the shape an application passes as `onEvent`. */
export class HandlerSink implements Sink {
  constructor(private readonly handler: RunEventHandler) {}

  onEvent(event: RunEvent): void {
    this.handler(event);
  }
}

/** Render an event as one human-readable line, mirroring the Python `format_event`. */
export function formatEvent(event: RunEvent): string {
  let line = `[${event.type}]`;
  if (event.step_id !== undefined) line += ` step=${event.step_id}`;
  if (event.message !== undefined) line += ` ${event.message}`;
  return line;
}
